### PR TITLE
feat(*): Implement the skeleton of an OTEL observability system

### PR DESCRIPTION
**Pull request by calebschoepp**
## Summary

This PR introduces OTEL observability into Spin — specifically tracing. Using the nomenclature of #2293 I have added some trigger observability while also cleaning up a bit of the runtime observability. This PR **DOES NOT** provide full tracing across the entirety of the Spin codebase. That would make for a huge PR so I opted to only trace a few parts and full tracing can be added in a followup PR.

The goal of this PR is to land in a place we're happy with the DX around observability leaving room for us to improve it in future PRs.

## Testing
```bash
# Install the otel-cli for easily producing distributed traces
go install github.com/equinix-labs/otel-cli@latest

# Run Jaeger
docker run -d -p16686:16686 -p4317:4317 -p4318:4318 -e COLLECTOR_OTLP_ENABLED=true jaegertracing/all-in-one:latest

# Run the app
OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318 spin up

# Call the app with the otel-cli
OTEL_EXPORTER_OTLP_ENDPOINT=localhost:4317 otel-cli exec -- bash -c 'curl localhost:3000-v -H "traceparent: $TRACEPARENT"'
```

### CODE REVIEW

File: crates/telemetry/src/config.rs
Comment (lann) [inline crates/telemetry/src/config.rs:14]: I still think this should be opt-in. See for example what containerd does: https://github.com/containerd/containerd/blob/9c77c389d897ffea482b2d56831b741aa0081eae/pkg/tracing/plugin/otlp.go#L131-L147

Comment (calebschoepp) [reply crates/telemetry/src/config.rs:14]: This is opt in. It's a weird double negative to turn it on disabled-false.

Comment (lann) [reply crates/telemetry/src/config.rs:14]: Ah that's...pretty weird too; sort of an "opt-out to not opting-in". That containerd approach seems better to me: an explicit otel endpoint configuration enables it but you can still override with `OTEL_SDK_DISABLED`

File: crates/telemetry/src/config.rs
Comment (calebschoepp) [inline crates/telemetry/src/config.rs:41]: Happy to bikeshed 🚲 🏠 this name.

File: crates/telemetry/src/config.rs
Comment (calebschoepp) [inline crates/telemetry/src/config.rs:34]: We don't technically use these config values other than to see if the user has set them so that we know if we should turn on an OTEL layer or not. `tracing-otlp` automatically reads these values from the environment.

File: crates/telemetry/src/config.rs
Comment (calebschoepp) [inline crates/telemetry/src/config.rs:75]: I could use help here. I was having trouble fighting the lifetimes.

File: crates/telemetry/src/traces.rs
Comment (calebschoepp) [inline crates/telemetry/src/traces.rs:86]: I still need to do these things before we can merge this.

File: crates/telemetry/src/config.rs
Comment (calebschoepp) [inline crates/telemetry/src/config.rs:17]: It occurs to me that we probably want the `service.name` to be whatever the user set as the app name in their manifest rather than just `Spin`. This would require us to read the manifest immediatley at the start of `spin.rs` since reloading the tracing subscriber isn't working. Are we okay with that?

Side note: How are we going to handle hot-swapping the OTEL resource in cloud?

Comment (lann) [reply crates/telemetry/src/config.rs:17]: Lets hold off on setting the service.name to the app name automatically until we understand the impact on shim users.

File: crates/telemetry/src/config.rs
Comment (lann) [reply crates/telemetry/src/config.rs:41]: We might want to prefix _this_ one with `SPIN_` since it isn't a standard OTEL env var.

File: crates/telemetry/src/config.rs
Comment (lann) [reply crates/telemetry/src/config.rs:34]: It might be less confusing to just list the vars that can enable tracing e.g.:

```rust
fn otlp_enabled() -> bool {
    const ENABLING_VARS: &[&str] = &[
        "OTEL_EXPORTER_OTLP_ENDPOINT",
        // ...
    ];
    ENABLING_VARS.iter().any(|key|
        std::env::var_os(key).is_some_and(|val| !val.is_empty()))
}
```

File: crates/telemetry/src/config.rs
Comment (lann) [reply crates/telemetry/src/config.rs:17]: Also I _think_ this can handle at least this var for you: https://docs.rs/opentelemetry_sdk/0.22.0/opentelemetry_sdk/resource/struct.EnvResourceDetector.html (with https://docs.rs/opentelemetry_sdk/0.22.0/opentelemetry_sdk/struct.Resource.html#method.from_detectors)

File: crates/telemetry/src/config.rs
Comment (lann) [reply crates/telemetry/src/config.rs:75]: Could you give more context? The `set_version` thing above seems fine to me...

File: crates/telemetry/src/traces.rs
Comment (lann) [inline crates/telemetry/src/traces.rs:60]: This is good context but it breaks up the flow of reading this function. I'd recommend extracting the mitigation (and comment) into a e.g. `fix_endpoint_bug(endpoint: &mut String)` function.

File: crates/telemetry/src/traces.rs
Comment (lann) [inline crates/telemetry/src/traces.rs:90]: Might be nice to split everything from here down into a e.g. `propagation` submodule. :shrug:

File: crates/trigger-http/src/handler.rs
Comment (lann) [inline crates/trigger-http/src/handler.rs:247]: Does this work properly when the debug level is disabled?

File: crates/telemetry/src/config.rs
Comment (rylev) [reply crates/telemetry/src/config.rs:75]: You'll have to copy the string data (using something like `to_owned`) - i.e., you can't directly store parts of `s` in a `KeyValues`.

File: src/bin/spin.rs
Comment (lann) [inline src/bin/spin.rs:52]: Maybe I'm just forgetting how we got here but do we still need the telemetry guard? Can we set the globals instead?

Comment (lann) [reply src/bin/spin.rs:52]: nm I forgot there is a global shutdown guard

File: crates/telemetry/src/config.rs
Comment (calebschoepp) [reply crates/telemetry/src/config.rs:41]: I had a similar thought, but remembered we inject into the environment variables at `SPIN_*` so I was concerned I was invading on that namespace. Perhaps it's not that big of a deal to steal the name `OTEL_TRACING_LEVEL` from that namespace.

File: crates/telemetry/src/config.rs
Comment (calebschoepp) [reply crates/telemetry/src/config.rs:34]: Yup, I like that more.

File: crates/trigger-http/src/handler.rs
Comment (calebschoepp) [reply crates/trigger-http/src/handler.rs:247]: Yes. The trace is still properly constructed in Jaeger even when the debug spans aren't set off to Jaeger.

File: crates/telemetry/src/config.rs
Comment (lann) [reply crates/telemetry/src/config.rs:41]: > inject into the environment variables at `SPIN_*`

Not sure what you mean here?

File: crates/trigger-http/src/handler.rs
Comment (lann) [reply crates/trigger-http/src/handler.rs:247]: I'm more concerned about `tracing` losing track of the "current" span if it drops out of scope before this spawned task completes. It would be good to verify with an app that does streaming responses like `examples/wasi-http-rust-streaming-outgoing-body`.

**Comment by calebschoepp** [comment]
@lann any ideas on how to test this feature?

@lann what are the steps to landing this. Any other stakeholders I should pull in for review other than you?

**Comment by lann** [comment]
> @lann any ideas on how to test this feature?

It would probably be fine to do most of the e2e testing in spinkube since most users of telemetry will be there in the medium term. For Spin itself I would probably just test that things don't entirely blow up when telemetry is 1) disabled, 2) enabled without debug traces, 3) enabled with debug traces.

**Comment by lann** [comment]
> > @lann what are the steps to landing this. Any other stakeholders I should pull in for review other than you?

I would suggest landing this asap after basic "things don't blow up" tests are in place. We'll want to iterate on user feedback.

File: crates/llm-local/src/lib.rs
Comment (rylev) [inline crates/llm-local/src/lib.rs:37]: I forget - by default what is the name of the span generated here? Do we want the spans for this to just say "infer"? Would be nice if it were a bit more clear from span name alone exactly what was happening. Something like "llm_generate_local_inference" and "llm_generate_local_embeddings".

**Comment by calebschoepp** [comment]
> > @lann any ideas on how to test this feature?
> 
> It would probably be fine to do most of the e2e testing in spinkube since most users of telemetry will be there in the medium term. For Spin itself I would probably just test that things don't entirely blow up when telemetry is 1) disabled, 2) enabled without debug traces, 3) enabled with debug traces.

I'll obviously test those cases in Spin locally. But, I'm wondering if we have any sort of e2e testing mechanism in Spin that I could use to automate the verification of it not blowing up.

File: crates/llm-local/src/lib.rs
Comment (rylev) [reply crates/llm-local/src/lib.rs:37]: Sounds good - as I've previously said, it would be amazing if we had some sort of system for deciding how to name spans. As we add more and more of these, we're going to simply be relying on author/reviewer judgement as to what a good span name is - it would be nicer if we gave some sort of guidance.

**Comment by rylev** [comment]
@calebschoepp We have [integration tests](https://github.com/fermyon/spin/blob/main/tests/integration.rs) that you might be able to cargo cult to fit your needs.

File: crates/telemetry/src/config.rs
Comment (lann) [reply crates/telemetry/src/config.rs:41]: Ohhh; the env var provider requires a prefix of `SPIN_VARIABLE_`, not just `SPIN_`. We use lots of vars that start with `SPIN_`.

**Comment by lann** [comment]
> @calebschoepp We have [integration tests](https://github.com/fermyon/spin/blob/main/tests/integration.rs) that you might be able to cargo cult to fit your needs.

Yeah you could even run jaeger as a testing-framework service and smoke test the actual span delivery.

File: crates/llm-local/src/lib.rs
Comment (calebschoepp) [reply crates/llm-local/src/lib.rs:37]: Names are updated.

To close the loop on the naming stuff: I'll flesh out the requirements for span names in my SIP.

File: crates/trigger-http/src/handler.rs
Comment (calebschoepp) [reply crates/trigger-http/src/handler.rs:247]: Karthik and I spent some time playing around with the examples to make sure this was working. Most of the examples were broken, but despite that we didn't observe any issues with tracing losing track of the "current" span. Let me know if you want further investigation here.

**Comment by calebschoepp** [comment]
@lann @rylev @itowlson this PR is in a place where I'm happy with it and it is ready for a final review.

File: crates/trigger-http/src/lib.rs
Comment (calebschoepp) [reply crates/trigger-http/src/lib.rs:352]: I was picturing this as a fast follow PR. I was struggling to get this to work and didn't want to block this PR on a few span attributes.

File: tests/testing-framework/services/jaeger.Dockerfile
Comment (calebschoepp) [reply tests/testing-framework/services/jaeger.Dockerfile:2]: In my experience it seems to start fairly quickly, but I have not quantitative evidence to back that up. My assumption was that the health checks came out of the box... What would I need to do to add a health check?

File: crates/telemetry/src/config.rs
Comment (itowlson) [inline crates/telemetry/src/config.rs:11]: I don't love the double negative here (or the "SDK") but I assume that this is forced on us by OpenTelemetry naming conventions?

File: crates/telemetry/src/config.rs
Comment (itowlson) [inline crates/telemetry/src/config.rs:15]: If it's Spin-specific, do we document the recognised values?  Should we reference that documentation here?  Could it be an enum?

File: crates/telemetry/src/config.rs
Comment (itowlson) [inline crates/telemetry/src/config.rs:21]: Is this name forced on us by OpenTelemetry, or could we rename it to something more specific such as `spin_version`?

File: crates/telemetry/src/config.rs
Comment (itowlson) [inline crates/telemetry/src/config.rs:37]: If I set one of these but also set `otel_sdk_disabled`, does the latter take precedence?

(ETA: reading elsewhere in the code, it appears yes.)

File: crates/telemetry/src/config.rs
Comment (itowlson) [inline crates/telemetry/src/config.rs:63]: Would the bit about defaults be more discoverable close to the `Config::otel_tracing_level` declaration?

File: crates/telemetry/src/detector.rs
Comment (itowlson) [inline crates/telemetry/src/detector.rs:25]: Is this the Spin version?

File: crates/telemetry/src/detector.rs
Comment (itowlson) [inline crates/telemetry/src/detector.rs:35]: This feels like a good candidate for being extracted to a variable and given a meaningful name.

File: crates/trigger-http/src/lib.rs
Comment (itowlson) [inline crates/trigger-http/src/lib.rs:342]: Do we have a convention on where these names come from?  I've seen a few go by and not been quite sure how they relate to function names or discrete activities within a function.  It's tricky I know because something like this is a lambda and has no name of its own...

File: crates/trigger-http/src/lib.rs
Comment (itowlson) [inline crates/trigger-http/src/lib.rs:356]: This 15-line code block dominates the lambda: it no longer stands out as "the service fn takes a request and handles it" but more "the service fn takes a request and does a metric butt-ton of tracing on it" - you have to read past a _lot_ of tracing fields to get to the actual behaviour.  Could this be encapsulated?

File: src/bin/spin.rs
Comment (itowlson) [inline src/bin/spin.rs:48]: This exits with success if it fails to load the telemetry config: this means that scripts that rely on the exit code will assume that Spin ran okay.  It should either continue without telemetry, or return an error (which Rust will turn into exit code 1).

**Review by itowlson — CHANGES_REQUESTED** [review]
90% of my comments are nits or sylistic fretting, and are non-blockers.  The one thing I'm concerned about is the error/exit behaviour in main.  If I've misread that (which is entirely likely!), let me know and I'll change to an approve!

Thank you for labouring through this.  It's a huge plumbing task, but super useful.

File: src/bin/spin.rs
Comment (itowlson) [reply src/bin/spin.rs:48]: Also, to print an error explicitly, prefer `terminal::error!` unless there is a compelling reason not to - this provides consistent formatting across the Spin UI.

File: crates/telemetry/src/config.rs
Comment (calebschoepp) [reply crates/telemetry/src/config.rs:11]: Correct, this is forced on us by OTEL [naming conventions](https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#general-sdk-configuration).

File: crates/telemetry/src/config.rs
Comment (calebschoepp) [reply crates/telemetry/src/config.rs:21]: The underlying thing this maps to is `system.version` which is forced on us by OTEL. But, this we could certainly give a nicer name like `spin_version`. Will do.

File: crates/telemetry/src/detector.rs
Comment (calebschoepp) [reply crates/telemetry/src/detector.rs:25]: Yep, I'll rename.

File: src/bin/spin.rs
Comment (calebschoepp) [reply src/bin/spin.rs:48]: This wasn't intentional to quell the error. I'll make the updates you suggested.

File: crates/trigger-http/src/lib.rs
Comment (calebschoepp) [reply crates/trigger-http/src/lib.rs:342]: Not really 🤷. Ryan has already raised the great point that we should have some kind of convention around these span names. Right now the best we have is something along the lines of:

> Make sure a span name can be understood without any additional surrounding context e.g. `generate_local_llm_inference` rather than `infer`.

I'm hoping to bake this into the SIP I wrote.

File: crates/telemetry/src/config.rs
Comment (calebschoepp) [reply crates/telemetry/src/config.rs:15]: Good call. I changed this to just directly use the underlying `tracing::Level` type that we parse it into.

File: src/bin/spin.rs
Comment (itowlson) [inline src/bin/spin.rs:47]: A shortcut if you want it:

```suggestion
            anyhow::bail!("Failed to load telemetry config: {err}"));
```

(definitely not important)

**Review by itowlson — APPROVED** [review]
Yay!

File: crates/telemetry/src/config.rs
Comment (lann) [inline crates/telemetry/src/config.rs:26]: I believe we could also use the same fmt filtering that `RUST_LOG` uses: https://docs.rs/tracing-subscriber/0.3.18/tracing_subscriber/filter/struct.EnvFilter.html

File: crates/telemetry/src/config.rs
Comment (lann) [inline crates/telemetry/src/config.rs:51]:
```suggestion
    /// Note that this is overridden if OTEL_SDK_DISABLED is set and not empty.
```

File: crates/telemetry/src/config.rs
Comment (lann) [inline crates/telemetry/src/config.rs:72]: If this is all we're using `envconfig` for I'd suggest dropping it and using `std::env::var` directly in `from_env`.

File: crates/telemetry/src/detector.rs
Comment (lann) [inline crates/telemetry/src/detector.rs:44]: It would be pretty weird for someone to set service.name to e.g. `MyApp` and for the version to be Spin's version, i.e. "MyApp version 0.1" would effectively look like "MyApp version 2.4.0" in spans.

Maybe we should be setting [`process.runtime.name` and `.version`](https://opentelemetry.io/docs/specs/semconv/attributes-registry/process/) for Spin instead?

File: crates/telemetry/src/lib.rs
Comment (lann) [inline crates/telemetry/src/lib.rs:65]: It doesn't look like the `Option<WorkerGuard>` is ever used?

File: crates/telemetry/src/detector.rs
Comment (lann) [inline crates/telemetry/src/detector.rs:30]: This seems to overlap with `EnvResourceDetector`. Is the point here just to set the default `service.name = spin`? If so, I suspect you could get the same effect with less code by manually creating a Resource in `otel_tracing_layer` and then [`merge`](https://docs.rs/opentelemetry_sdk/0.22.1/opentelemetry_sdk/struct.Resource.html#method.merge)ing in the detected resource attrs.

File: src/bin/spin.rs
Comment (lann) [reply src/bin/spin.rs:47]: Also, would `.context("Failed to load telemetry config")` be appropriate here?

File: tests/testing-framework/services/jaeger.Dockerfile
Comment (lann) [reply tests/testing-framework/services/jaeger.Dockerfile:2]: Something like this:

```
HEALTHCHECK --interval=2s CMD wget -qO - http://localhost:14269/health
```

File: crates/telemetry/src/config.rs
Comment (calebschoepp) [reply crates/telemetry/src/config.rs:26]: We could certainly use the EnvFilter if we wanted. I opted for the LevelFilter b/c I wanted to expose less knobs to the end user. But, my whole note there is basically saying "hey eventually we'll want to not just expose LevelFilter but actually the whole EnvFilter". So given that EnvFilter is capable of acting like a LevelFilter too I think you're right that is best to just switch to using an EnvFilter.

File: crates/telemetry/src/detector.rs
Comment (calebschoepp) [reply crates/telemetry/src/detector.rs:44]: Agreed about that being weird. But, the user can also set the version if they're setting a custom name e.g. `OTEL_RESOURCE_ATTRIBUTES=service.version=78 OTEL_SERVICE_NAME=mycustomname`. I don't think it is un-reasonable to expect them to do that if they want a custom name. I also think it is important to provide the Spin version out of the box.

I think we should switch to putting this info in `process.runtime.{name, version}` like you suggest in a future PR when we start supporting automatically setting the `service.name` based off the manifest application name. Then we can also automatically set `service.version` based off the manifest application version.

File: crates/telemetry/src/detector.rs
Comment (calebschoepp) [reply crates/telemetry/src/detector.rs:30]: It also sets the version.

That's a good idea, but I think I prefer keeping it in the detector. Inlining it would remove the need for a struct definition and a constructor, but otherwise most of the code would remain. I prefer following the pattern of a detector.

Comment (lann) [reply crates/telemetry/src/detector.rs:30]: That's fine - and this certainly isn't a blocking comment - but if its effectively setting a default and a hard-coded version I might be tempted to argue that it isn't "detecting" anything. :slightly_smiling_face:

File: crates/trigger-http/src/lib.rs
Comment (calebschoepp) [reply crates/trigger-http/src/lib.rs:356]: I'm sure it can, but I'm scared to touch code that is this critical to Spin right before I merge this. I'll make this stylistic change in a fast-follow PR when I come back to add these attributes that are added later.

Comment (lann) [reply crates/trigger-http/src/lib.rs:356]: If you extracted it into a function then the call site information would be wrong. You could extract it into a macro instead if you want.

Comment (lann) [reply crates/trigger-http/src/lib.rs:356]: As a further tangent, I just learned that you can use constants as the field names, such as those from [`opentelemetry-semantic-conventions`](https://docs.rs/opentelemetry-semantic-conventions/latest/opentelemetry_semantic_conventions/trace/index.html); `tracing` just requires that you wrap constants in `{}`, e.g.:

```rust
mod keep_all_these_constants_hidden_away {
  use opentelemetry_semantic_conventions::trace::*;
  ...
    span!(
      ...
      { HTTP_REQUEST_METHOD } = method,
    )
}
```

Comment (lann) [reply crates/trigger-http/src/lib.rs:356]: (but agreed this can come in a followup, if at all)

**Pull request merged**
